File: DataStructures/Tree/SegmentationTree/python/SegmentationTree.py
# Sum of given range implementation

import sys
import math

def updateUtil(st, s, e, i, diff, si):
	if(i < s or i > e):
		return
	st[si] += diff
	if(s != e):
		mid = s + ((e - s) // 2)
		updateUtil(st, s, mid, i, diff, 2*si + 1)
		updateUtil(st, mid+1, e, i, diff, 2*si + 2)

def update(st,  a,  n, index, newValue):
	if(index < 0 or index > n-1):
		print("Invalid input", end = "\n")
		return

	diff = newValue - a[index]
	a[index] = newValue

	updateUtil(st, 0, n-1, index, diff, 0)

def getSumUtil(st, s, e, l, r, si):
	if(s > e or l > e or r < s):
		return 0
	if(s >= l and e <= r):
		return st[si]
	mid = s + ((e - s) // 2)
	return getSumUtil(st, s, mid, l, r, 2*si + 1) + getSumUtil(st, mid + 1, e, l, r, 2*si + 2)

def getSum(st, n, l, r):
	if(l < 0 or r > n-1 or l > r):
		print("Invalid input: ")
		return -1

	print("sum in the range (",l,", ",r,") is ")
	return getSumUtil(st, 0, n-1, l, r, 0)

def buildSTutil(st, a, s, e, si):
	if(s == e):
		st[si] = a[s]
		return a[s]

	mid = s + ((e - s) // 2)

	st[si] = buildSTutil(st, a, s, mid, 2*si+1) + buildSTutil(st, a, mid+1, e, 2*si+2)
	return st[si]

def buildST(a, n):
	height = (int)(math.ceil(math.log2(n)))
	maxSize = 2*(int)(2**height) - 1

	st = [0] * maxSize
	buildSTutil(st, a, 0, n-1, 0)

	return st

if __name__ == "__main__":

	sys.stdin = open('../input.txt', 'r')
	sys.stdout = open('../output.txt', 'w')

	n = int(sys.stdin.readline())
	a = sys.stdin.readline().split(" ") # [:-1] to eliminate reading of newline
	a = [int(i) for i in a]

	st = buildST(a, n)

	for line in sys.stdin:

		if line[-1] == '\n':
			opt = line[:-1].split(" ")
		else:
			opt = line.split(" ")
		
		if opt[0] == "sum":
			l, r = int(opt[1]), int(opt[2])
			sum_ = getSum(st, n, l, r)
			print(sum_, end = "\n");
		elif opt[0] == "update":
			index, value = int(opt[1]), int(opt[2])
			update(st, a, n, index, value)

File: DataStructures/Tree/SegmentationTree/c/SegmentationTree.c
// Sum of Given range Implementation

// run: gcc Segmentation.c -lm
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdbool.h>
#include <math.h>

void updateUtil(int st[], int s, int e, int i, int diff, int si){
	if(i < s || i > e)
		return;
	st[si]  += diff;
	if(s != e){
		int mid = s + ((e - s) / 2);
		updateUtil(st, s, mid, i, diff, 2*si + 1);
		updateUtil(st, mid+1, e, i, diff, 2*si + 2);
	}
}
void update(int st[], int a[], int n, int index, int newValue){
	if(index < 0 || index > n-1){
		printf("Invalid input\n");
		return;
	}
	int diff = newValue - a[index];
	a[index] = newValue;

	updateUtil(st, 0, n-1, index, diff, 0);
}

int getSumUtil(int st[], int s, int e, int l, int r, int si){
	if(s > e || l > e || r < s)
		return 0;
	if(s >= l && e <= r)
		return st[si];
	int mid = s + ((e - s) / 2);
	return getSumUtil(st, s, mid, l, r, 2*si + 1) + getSumUtil(st, mid + 1, e, l, r, 2*si + 2);
}
int getSum(int st[], int n, int l, int r){
	if(l < 0 || r > n-1 || l > r){
		printf("Invalid input: ");
		return -1;
	}

	printf("sum in the range (%d, %d) is ", l, r);
	return getSumUtil(st, 0, n-1, l, r, 0);
}

int buildSTutil(int st[], int a[], int s, int e, int si){
	if(s == e){
		st[si] = a[s];
		return a[s];
	}
	int mid = s + ((e - s) / 2);

	st[si] = buildSTutil(st, a, s, mid, 2*si+1) + buildSTutil(st, a, mid+1, e, 2*si+2);
	return st[si];
}

int main(){

	FILE* fp = freopen("../input.txt", "r", stdin);
	freopen("../output.txt", "w", stdout);

	int i,n, a[n];
	scanf("%d", &n);
	for(i=0; i<n; i++)
		scanf("%d", &a[i]);

	int height = (int)(ceil(log(n)/log(2)));

	int maxSize = 2*(int)(pow(2, height)) - 1;

	int st[maxSize];
	int ret = buildSTutil(st, a, 0, n-1, 0);

	while(!feof(fp)){

		char opt[20]; 
		fgets(opt, 20, fp); // it reads string from a file, one line at a time

		if(!strncmp(opt,"sum",strlen("sum"))){ 
			char choice[10];
			int l,r;
			sscanf(opt, "%s %d %d" , choice, &l, &r); //sscanf reads formatted input from string

			int sum = getSum(st, n, l, r);
			printf("%d\n", sum); 
		}
		else if(!strncmp(opt,"update",strlen("update"))){
			char choice[10];
			int index, value;
			sscanf(opt, "%s %d %d" , choice, &index, &value);

			update(st, a, n, index, value); 
		}
	}
	fclose(fp); 
}